components/core/src/metrics.rs

```
@@ -0,0 +1,221 @@
+    let (rztx, rzrx) = sync_channel(0); // rendezvous channel
+
+    thread::Builder::new()
+        .name("metrics".to_string())
+        .spawn(move || receive(rztx, rx))
+        .expect("couldn't start metrics thread");
+
+    match rzrx.recv() {
+        Ok(()) => tx,
+        Err(e) => panic!("metrics thread startup error, err={}", e),
+    }
+}
+
+// receive runs in a separate thread and processes all metrics events
+fn receive(rz: SyncSender<()>, rx: Receiver<MetricTuple>) {
+    let mut client = statsd_client();
+    rz.send(()).unwrap(); // Blocks until the matching receive is called
+
+    loop {
+        let (mtyp, mop, mid, mval): MetricTuple = rx.recv().unwrap();
+        debug!("Received metrics tuple: {:?}", (mtyp, mop, mid, mval));
+
+        match client {
+            Some(ref mut cli) => {
+                match mtyp {
+                    MetricType::Counter => {
+                        match mop {
+                            MetricOperation::Increment => cli.incr(mid),
+                            MetricOperation::Decrement => cli.decr(mid),
+                            _ => error!("Unexpected metric operation: {:?}", mop),
+                        }
+                    }
+                    MetricType::Gauge => {
+                        match mop {
+                            MetricOperation::SetValue => cli.gauge(mid, mval.unwrap()),
+                            _ => error!("Unexpected metric operation: {:?}", mop),
+                        }
+                    }
+                }
+            }
+            None => (),
+        }
+    }
+}
+
+fn statsd_client() -> Option<Client> {
+    match env::var(STATS_ENV) {
+        Ok(addr) => Some(Client::new(&addr, APP_NAME).unwrap()),
+        Err(_) => None,
+    }
+}
+
+impl Counter {
+    pub fn increment(&self) {
+        sender()
+            .send((MetricType::Counter, MetricOperation::Increment, &self.id(), None))
+            .unwrap();
```

Probably better to return an error unless the caller can't easily cope with failures--I'm not sure.

@adamhjk I've been wanting to have a chat with the team about `.expect()`. While much, much nicer and more expressive that `.unwrap()`, it still causes a panic on `Err`/`None`. Looking back through some of our code in the last few months I'm concerned that we've introduced more places which can terminate a thread of execution.